### DISCUSSION TITLE
How to use ReadLine activity in dashboard?

### DISCUSSION THREAD

**Discussion by maiorfi** [elsa-workflows/elsa-core#2778] [Q&A]
How is stdin accessed, actually?

**Comment by sfmskywalker** [comment]
By default, it's accessing stdin via `Console.In`. Currently there's no easy way to register this activity with a custom `TextReader` without registing `TextReader` with DI, but if this is something you're interested in I'd be OK with refactoring the `ReadLine` activity so that you could provide access to stdin via a servce for example.

As far as I know, ReadLine is only really used in some of the console sample apps.

**Reply by maiorfi** [reply]
So it's not possible to run an instance of those workflows via dashboard and provide some sample console input either, right?

I was wondering how to test workflow just edited/authored interactively in dashboard.

Thanks.

**Reply by sfmskywalker** [reply]
Exactly, the dashboard doesn't currently provide a tunnel to the console input. Maybe this is something we could do for V3 as we expand on testing capabilities where we want to stream console out to the browser. We might as well provide an input terminal.

Regardless of lack of support for console in, there is (rudimentary) support for running from the dashboard that you can enable. To enable this feature:

1. Reference the `Elsa.WorkflowTesting.Api` package
2. Add this to Startup: `services.AddWorkflowTestingServices();`
3. Add this to Startup: `app.UseEndpoints(endpoints => endpoints.MapWorkflowTestHub());`

Then you will be able to run a workflow from the dashboard, like this:

![test-workflow-from-dashboard](https://user-images.githubusercontent.com/938393/158336184-e08da0e6-2a2f-4e3b-9e76-d7227c998aaf.gif)

In future iterations, we want to add the ability to set breakpoints on individual activities to allow for introspection of current activity state.

**Reply by maiorfi** [reply]
Very cool! Thanks.
